bin/runtime-common/src/messages.rs

```diff
@@ -0,0 +1,539 @@
+			if actual_fee_in_bridged_tokens < minimal_fee_in_bridged_tokens {
+				return Err("Too low fee paid");
+			}
+
+			Ok(())
+		}
+	}
+
+	/// Estimate delivery and dispatch fee that must be paid for delivering a message to the Bridged chain.
+	///
+	/// The fee is paid in This chain Balance, but we use Bridged chain balance to avoid additional conversions.
+	/// Returns `None` if overflow has happened.
+	pub fn estimate_message_dispatch_and_delivery_fee<B: MessageBridge>(
+		payload: &FromThisChainMessagePayload<B>,
+		relayer_fee_percent: u32,
+	) -> Result<BalanceOf<BridgedChain<B>>, &'static str> {
+		// the fee (in Bridged tokens) of all transactions that are made on the Bridged chain
+		let delivery_fee = B::bridged_weight_to_balance(B::weight_of_delivery_transaction());
+		let dispatch_fee = B::bridged_weight_to_balance(payload.weight.into());
+		let reward_confirmation_fee =
+			B::bridged_weight_to_balance(B::weight_of_reward_confirmation_transaction_on_target_chain());
+
+		// the fee (in Bridged tokens) of all transactions that are made on This chain
+		let delivery_confirmation_fee = B::this_chain_balance_to_bridged_chain_balance(B::this_weight_to_balance(
+			B::weight_of_delivery_confirmation_transaction_on_this_chain(),
+		));
+
+		// minimal fee (in Bridged tokens) is a sum of all required fees
+		let minimal_fee = delivery_fee
+			.checked_add(&dispatch_fee)
+			.and_then(|fee| fee.checked_add(&reward_confirmation_fee))
+			.and_then(|fee| fee.checked_add(&delivery_confirmation_fee));
+
+		// before returning, add extra fee that is paid to the relayer (relayer interest)
+		minimal_fee
+			.and_then(|fee|
+			// having message with fee that is near the `Balance::MAX_VALUE` of the chain is
+			// unlikely and should be treated as an error
+			// => let's do multiplication first
+			fee
+				.checked_mul(&relayer_fee_percent.into())
+				.and_then(|interest| interest.checked_div(&100u32.into()))
+				.and_then(|interest| fee.checked_add(&interest)))
+			.ok_or("Overflow when computing minimal required message delivery and dispatch fee")
+	}
+}
+
+/// Sub-module that is declaring types required for processing Bridged -> This chain messages.
+pub mod target {
+	use super::*;
+
+	/// Call origin for Bridged -> This chain messages.
+	pub type FromBridgedChainMessageCallOrigin<B> = pallet_bridge_call_dispatch::CallOrigin<
+		SignerOf<BridgedChain<B>>,
+		SignerOf<ThisChain<B>>,
+		SignatureOf<ThisChain<B>>,
+	>;
+
+	/// Decoded Bridged -> This message payload.
+	pub type FromBridgedChainDecodedMessagePayload<B> = pallet_bridge_call_dispatch::MessagePayload<
+		SignerOf<BridgedChain<B>>,
+		SignerOf<ThisChain<B>>,
+		SignatureOf<ThisChain<B>>,
+		CallOf<ThisChain<B>>,
+	>;
+
+	/// Message payload for Bridged -> This messages.
+	pub struct FromBridgedChainMessagePayload<B: MessageBridge>(pub(crate) FromBridgedChainDecodedMessagePayload<B>);
+
+	impl<B: MessageBridge> Decode for FromBridgedChainMessagePayload<B> {
```

Comment:
Correct. This custom `Decode` impl should yet to be tested in real environment, though (after we'll have everything else ready). The assumption is that:
1) the sender would submit tx with `BridgedChain::Call::encode()`, because `message-lane` module expects `Vec<u8>`;
2) when message payload will be encoded by `message-lane` module, it'll be prefixed with vector lenght;
3) so on receiving side we'll need to skip that length and proceed directly to wrapped encoded `Call`